File: examples/injectable-app/java/src/main/java/com/twitter/finatra/example/HelloWorldApp.java
package com.twitter.finatra.example;

import java.util.Collection;
import java.util.Collections;
import java.util.List;

import com.google.inject.Module;

import com.twitter.app.Flaggable;
import com.twitter.inject.annotations.Flags;
import com.twitter.inject.app.AbstractApp;
import com.twitter.inject.modules.StatsReceiverModule;
import com.twitter.util.logging.Logger;

public class HelloWorldApp extends AbstractApp {
  private static final Logger LOG = Logger.getLogger(HelloWorldApp.class);

  private final List<Integer> queue;

  public HelloWorldApp(List<Integer> q) {
    this.queue = q;

    createMandatoryFlag(
        "username",
        "Username to use.",
        "-username=Bob",
        Flaggable.ofString()
    );
  }

  @Override
  public Collection<Module> javaModules() {
    return Collections.singletonList(StatsReceiverModule.get());
  }

  @Override
  public void run() {
    queue.add(3);
    HelloService helloService = injector().instance(HelloService.class);

    // username Flag is mandatory. if it has no value, the app fails here.
    String username = injector().instance(String.class, Flags.named("username"));
    LOG.debug(String.format("Input username: %s", username));
    LOG.info(helloService.hi(username));
  }

  @Override
  public void onInit() {
    queue.add(1);
  }

  @Override
  public void preMain() {
    queue.add(2);
  }

  @Override
  public void postMain() {
    queue.add(4);
  }

  @Override
  public void onExit() {
    queue.add(5);
  }

  @Override
  public void onExitLast() {
    queue.add(6);
  }

  public List<Integer> getQueue() {
    return queue;
  }
}

File: examples/injectable-app/scala/src/main/scala/com/twitter/finatra/example/HelloWorldApp.scala
package com.twitter.finatra.example

import com.google.inject.Module
import com.twitter.inject.annotations.Flags
import com.twitter.inject.app.App
import com.twitter.inject.modules.StatsReceiverModule
import com.twitter.util.logging.Logger
import com.twitter.util.logging.Slf4jBridge
import scala.collection.mutable.ArrayBuffer

object SampleAppMain extends SampleApp

class SampleApp extends App with Slf4jBridge {
  private[this] val log: Logger = Logger("SampleApp")

  private[this] val queue: ArrayBuffer[Int] = new ArrayBuffer[Int]()

  flag[String]("username", "Username to use.", "-username=Bob")

  override val modules: Seq[Module] = Seq(StatsReceiverModule)

  override protected def run(): Unit = {
    queue += 3
    val helloService: HelloService = injector.instance[HelloService]

    // username Flag is mandatory. if it has no value, the app fails here.
    val username: String = injector.instance[String](Flags.named("username"))
    log.debug(s"Input username: $username")
    log.info(helloService.hi(username))
  }

  init {
    queue += 1
  }

  premain {
    queue += 2
  }

  postmain {
    queue += 4
  }

  onExit {
    queue += 5
  }

  onExitLast {
    queue += 6
  }

  def getQueue: Seq[Int] = this.queue.toSeq
}
